src/reducers/windowHandler.js

```diff
@@ -238,33 +243,42 @@ export default function windowHandler(state = initialState, action) {
         },
       });
     case ADD_ROW_DATA:
-      return Object.assign({}, state, {
-        [action.scope]: Object.assign({}, state[action.scope], {
-          rowData: Object.assign({}, state[action.scope].rowData, action.data),
-        }),
+      let addRowData = Map();
+
+      map(action.data, (item, key) => {
+        const arrayItem = item.length ? item : [];
+        addRowData = addRowData.set(key, List(arrayItem));
       });
-    case ADD_NEW_ROW:
-      return update(state, {
+
+      return {
+        ...state,
         [action.scope]: {
-          rowData: {
-            [action.tabid]: {
-              $push: [action.item],
-            },
-          },
+          ...state[action.scope],
+          rowData: state[action.scope].rowData.merge(addRowData),
         },
-      });
+      };
+    case ADD_NEW_ROW:
+      const newRowData = state[action.scope].rowData.update(`${action.tabid}`, list => list.push(action.item));
+
+      return {
+        ...state,
+        [`${action.scope}`]: {
+          ...state[`${action.scope}`],
+          rowData: newRowData,
+        },
+      };
     case DELETE_ROW:
-      return update(state, {
-        [action.scope]: {
-          rowData: {
-            [action.tabid]: {
-              $set: state[action.scope].rowData[action.tabid].filter(
-                item => item.rowId !== action.rowid
-              ),
-            },
-          },
+      const deletedRowData = state[action.scope].rowData.update(`${action.tabid}`,
+        list => list.filter(item => item.rowId !== action.rowid)
+      );
+
+      return {
+        ...state,
+        [`${action.scope}`]: {
+          ...state[`${action.scope}`],
+          rowData: deletedRowData,
         },
-      });
+      };
```

Comment:
To benefit from `Immutable.js` the state object itself should be an immutable data structure and you should remove plain JS destructuring across the board. To persistently update deeply nested values, use https://facebook.github.io/immutable-js/docs/#/Map/setIn et al.

In this case it should be something like:

```js
state.updateIn(
    [action.scope, 'rowDate', action.tabid],
    state
        .getIn([action.scope, 'rowDate', action.tabId])
        .filter(item => item.rowId !== action.rowid)
);
```

That way we can simply check for reference equality, if anything changed on the state. (So it won't change for example if we didn't actually filter out any value)

Reply:
The problem is, changing the whole state to Immutable type might cause a lot of issues. I think we should to it gradually. My motivation here was to avoid expensive deep equal checks when rendering components. And table rows are most prone to this, since we can have structures nested on multiple levels.
But I get your point :) I'd love to have everything immutable + type checks. I hope we'll get there at some point. Rolling out flow is high on my priority list.

Reply:
Generally, I agree with you on the gradual changes.

However, if we take the right approach now and issues arise, we are actually fixing the problem instead of patching around it, since we get the foundation right. In some cases we would benefit greatly if we ripped out the components completely and rewrote them.

If we look at the surface, we are moving in the right direction - however, the components are still way too coupled to implementation details instead of providing a generic interface. It's the same thing that prevents us from writing sensible regression tests / unit tests. Meaning that sooner of later we'll get bitten again.

Be careful about the type checks - they are not solving an inherent problem we have right now. When you introduce a type system, you are hardening existing structures. That means that we'll reinforce the coupling that's existent in many part of the apps and will make refactorings even harder.

Using types makes sense if you are confident that the specification of the code you wrote won't change significantly in the future. That will most likely only apply to new components that we write.

I'd really like to hear your case on what would justify introducing types to the current state of the code base

Reply:
hi @PabloSichert 
here is my opinion from a non js-dev perspective:

> However, if we take the right approach now and issues arise [...]

...if they pop up in `master` could they be fixed until the resp. next friday? Or could they all be identified and fixed in the respective refacture-feature-branch while also constantly adding unrelated fixes/features to `master` ?

Unless I'm convinced that the answer is "yes, and btw, we will also get new features done in parallel", I would prefer things to continue the way @siemiatj does them.

Reply:
agree. we have some experience in the past with approaches and i can tell that the mozilla way is not for everybody :)

Reply:
Fair enough 😄

I didn't mean to just rip out the code, push it to production and see what happens, though.

We can take a look at pieces of the code that would warrant a bigger refactoring and plan it in advance. In my opinion the List / Dropdown things are candidates for that. To accomplish that with a same process, it would be nice build specific branches to a test server and have them go through QA before we land it on master. @metas-ts Is there already some infrastructure in place for that?

I just want to have components generic enough that testing and regression tests are straight forward.

Let's take the Shortcuts component as an example. It took some time to refactor, but now they are generic, tested and don't do any issues any more that we couldn't tackle in a regression test.

I'm not sure, guys - but aren't you as tired as me of the same issues popping up again over and over?